# ETL - Siniestros Viales

Importamos las librerias que utilizaremos en esta etapa del proyecto.

In [915]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 


## Extracción de Datos:

Comenzamos cargando el archivo Excel que contiene dos hojas para luego pasar cada una de ellas a un csv y así trabajamos con ellas con mayor facilidad.

In [916]:
excel_file = "homicidios.xlsx"

hechos = pd.read_excel(excel_file, sheet_name="HECHOS")
victimas = pd.read_excel(excel_file, sheet_name="VICTIMAS")

hechos.to_csv("hechos.csv", index=False)
victimas.to_csv("victimas.csv", index=False)

**DataFrame**
Para trabajar de una manera más ordenada, realizamos el ETL de ambos csv a la par. Creamos por consiguiente el DataFrame de "hechos" y el DataFrame de "victimas".

In [917]:
df_hechos = pd.read_csv("hechos.csv", encoding='UTF-8')
display(df_hechos.head())


ID  N_VICTIMAS       FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1  2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1  2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1  2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1  2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1  2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

In [918]:
df_victimas = pd.read_csv("victimas.csv", encoding='utf-8')
display(df_victimas.head())

ID_hecho       FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO EDAD  \
0  2016-0001  2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002  2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003  2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004  2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005  2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

   FECHA_FALLECIMIENTO  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

In [919]:
df_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ID                     696 non-null    object 
 1   N_VICTIMAS             696 non-null    int64  
 2   FECHA                  696 non-null    object 
 3   AAAA                   696 non-null    int64  
 4   MM                     696 non-null    int64  
 5   DD                     696 non-null    int64  
 6   HORA                   696 non-null    object 
 7   HH                     696 non-null    object 
 8   LUGAR_DEL_HECHO        696 non-null    object 
 9   TIPO_DE_CALLE          696 non-null    object 
 10  Calle                  695 non-null    object 
 11  Altura                 129 non-null    float64
 12  Cruce                  525 non-null    object 
 13  Dirección Normalizada  688 non-null    object 
 14  COMUNA                 696 non-null    int64  
 15  XY (CA

In [920]:
df_victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   ID_hecho             717 non-null    object
 1   FECHA                717 non-null    object
 2   AAAA                 717 non-null    int64 
 3   MM                   717 non-null    int64 
 4   DD                   717 non-null    int64 
 5   ROL                  717 non-null    object
 6   VICTIMA              717 non-null    object
 7   SEXO                 717 non-null    object
 8   EDAD                 717 non-null    object
 9   FECHA_FALLECIMIENTO  717 non-null    object
dtypes: int64(3), object(7)
memory usage: 56.1+ KB


## Transformación de Datos:

Continuamos con la transformación. Esta consiste en limpiar, modificar y preparar los datos para su uso posterior. Eso incluye eliminar datos nulos, convertir tipos de datos, normalizar columnas, desanidar listas, etc.

### DATOS NULOS y COLUMNAS IRRELEVANTES
En primer lugar voy a chequear si hay o no **id** que tengan **valor nulo**, dado que eso me podria traer problemas al relacionar mis datos con los otros dataset. De existir **id** sin valor, entonces eliminaría esas filas.

**Hechos**

In [921]:
# Verificar valores nulos en la columna 'ID'
id_null = df_hechos['ID'].isnull().sum()
print(f'El dataset tiene {id_null} filas sin el dato ID.')

# Si hay filas con valores nulos en 'ID', eliminarlas; de lo contrario, conservar el DataFrame original
if id_null > 0:
    df_hechos = df_hechos.dropna(subset=['ID'])
    print(f'El nuevo dataset de hechos tiene {df_hechos.shape[0]} filas.')
else:
    print('No hay filas en hechos con valores nulos en la columna ID. El DataFrame original se conserva.')


El dataset tiene 0 filas sin el dato ID.
No hay filas en hechos con valores nulos en la columna ID. El DataFrame original se conserva.


**Victimas**

In [922]:
# Verificar valores nulos en la columna 'ID'
id_null_v = df_victimas['ID_hecho'].isnull().sum()
print(f'El dataset tiene {id_null_v} filas sin el dato ID.')

# Si hay filas con valores nulos en 'ID', eliminarlas; de lo contrario, conservar el DataFrame original
if id_null_v > 0:
    df_victimas = df_victimas.dropna(subset=['ID_hecho'])
    print(f'El nuevo dataset de victimas tiene {df_victimas.shape[0]} filas.')
else:
    print('No hay filas en victimas con valores nulos en la columna ID. El DataFrame original se conserva.')


El dataset tiene 0 filas sin el dato ID.
No hay filas en victimas con valores nulos en la columna ID. El DataFrame original se conserva.


Ahora analizamos en qué columnas hay datos nulos y de cuantos se tratan.

**Hechos**

In [923]:
nulos = df_hechos.isnull().sum()
print(nulos)


ID                         0
N_VICTIMAS                 0
FECHA                      0
AAAA                       0
MM                         0
DD                         0
HORA                       0
HH                         0
LUGAR_DEL_HECHO            0
TIPO_DE_CALLE              0
Calle                      1
Altura                   567
Cruce                    171
Dirección Normalizada      8
COMUNA                     0
XY (CABA)                  0
pos x                      0
pos y                      0
PARTICIPANTES              0
VICTIMA                    0
ACUSADO                    0
dtype: int64


Los valores nulos son significativos en el df_hechos en dos columnas en particular: Altura con 567 valores nulos y Cruce con 171 nulos.

**Victimas**

In [924]:
nulos_v = df_victimas.isnull().sum()
print(nulos_v)

ID_hecho               0
FECHA                  0
AAAA                   0
MM                     0
DD                     0
ROL                    0
VICTIMA                0
SEXO                   0
EDAD                   0
FECHA_FALLECIMIENTO    0
dtype: int64


En el caso de victimas no hay datos nulos.

Ahora analizamos en qué columnas se registro "SD", es decir, sin datos, y de cuantos se tratan. Notamos que puede haber cierta variación en columnas como PARTICIPANTES dado que muestra cero pero sabemos que probablemente se deba a que tiene el dato de alguno de los integrantes.

**Hechos**

In [925]:
sin_datos = (df_hechos == 'SD').sum()
print(sin_datos)

ID                        0
N_VICTIMAS                0
FECHA                     0
AAAA                      0
MM                        0
DD                        0
HORA                      1
HH                        1
LUGAR_DEL_HECHO           1
TIPO_DE_CALLE             0
Calle                     0
Altura                    0
Cruce                     0
Dirección Normalizada     0
COMUNA                    0
XY (CABA)                 0
pos x                     0
pos y                     0
PARTICIPANTES             0
VICTIMA                   9
ACUSADO                  23
dtype: int64


**Victimas**

In [926]:
sin_datos_v = (df_victimas == 'SD').sum()
print(sin_datos_v)

ID_hecho                0
FECHA                   0
AAAA                    0
MM                      0
DD                      0
ROL                    11
VICTIMA                 9
SEXO                    6
EDAD                   53
FECHA_FALLECIMIENTO    68
dtype: int64


### Eliminación de columnas.

**Hechos**

Decidimos **eliminar** las columnas **Altura, Hora, XY (CABA) y Dirección Normalizada** para simplificar nuestro dataframe y evitar redundancias en los datos de ubicación. Elegimos por mantener solo aquellas columnas que proporcionan información más precisa y útil para nuestro análisis y visualización en Power BI. De esta manera, nos enfocamos en datos clave que faciliten un análisis más claro y efectivo.

In [927]:
columnas_destacadas = ['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HH', 'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce', 'COMUNA', 'pos x', 'pos y', 'PARTICIPANTES', 'VICTIMA', 'ACUSADO']
df_hechos2 = df_hechos[columnas_destacadas]

print(df_hechos2.head())

          ID  N_VICTIMAS       FECHA  AAAA  MM  DD HH  \
0  2016-0001           1  2016-01-01  2016   1   1  4   
1  2016-0002           1  2016-01-02  2016   1   2  1   
2  2016-0003           1  2016-01-03  2016   1   3  7   
3  2016-0004           1  2016-01-10  2016   1  10  0   
4  2016-0005           1  2016-01-21  2016   1  21  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE             Calle  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  PIEDRA BUENA AV.   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ    PAZ, GRAL. AV.   
2                         AV ENTRE RIOS 2034       AVENIDA    ENTRE RIOS AV.   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA    LARRAZABAL AV.   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA      SAN JUAN AV.   

                                 Cruce  COMUNA         pos x         pos y  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.       8  -58.47533969  -34.68757022   
1                  D

In [928]:
df_hechos2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ID               696 non-null    object
 1   N_VICTIMAS       696 non-null    int64 
 2   FECHA            696 non-null    object
 3   AAAA             696 non-null    int64 
 4   MM               696 non-null    int64 
 5   DD               696 non-null    int64 
 6   HH               696 non-null    object
 7   LUGAR_DEL_HECHO  696 non-null    object
 8   TIPO_DE_CALLE    696 non-null    object
 9   Calle            695 non-null    object
 10  Cruce            525 non-null    object
 11  COMUNA           696 non-null    int64 
 12  pos x            696 non-null    object
 13  pos y            696 non-null    object
 14  PARTICIPANTES    696 non-null    object
 15  VICTIMA          696 non-null    object
 16  ACUSADO          696 non-null    object
dtypes: int64(5), object(12)
memory usag

**Victimas**

Hay datos que se repiten de la tabla anterior con lo cual no son necesarios conservar

In [929]:
df_victimas2 = df_victimas.drop(columns=['FECHA', 'AAAA', 'MM', 'DD', 'VICTIMA', 'FECHA_FALLECIMIENTO'])
df_victimas2

ID_hecho                   ROL       SEXO EDAD
0    2016-0001             CONDUCTOR  MASCULINO   19
1    2016-0002             CONDUCTOR  MASCULINO   70
2    2016-0003             CONDUCTOR  MASCULINO   30
3    2016-0004             CONDUCTOR  MASCULINO   18
4    2016-0005             CONDUCTOR  MASCULINO   29
..         ...                   ...        ...  ...
712  2021-0092                PEATON   FEMENINO   50
713  2021-0093  PASAJERO_ACOMPAÑANTE   FEMENINO   18
714  2021-0094  PASAJERO_ACOMPAÑANTE   FEMENINO   43
715  2021-0095             CONDUCTOR  MASCULINO   27
716  2021-0096             CONDUCTOR  MASCULINO   60

[717 rows x 4 columns]

### DATOS DUPLICADOS

A continuación revisamos si hay **datos duplicados**. Empezaremos por los id en el csv hechos, dado que deben ser únicos. Y a continuación chequeamos si hay filas que aún con id diferente comparten misma data en todo el resto de sus campos (eso podria deberse a error humano de dobles cargas por ejemplo)

**Hechos**

In [930]:
# Verifico ID duplicados
ids_duplicados = df_hechos2['ID'].duplicated().sum()

if ids_duplicados > 0:
    print(f"Hay {ids_duplicados} IDs duplicados.")
else:
    print("No hay IDs duplicados.")

# Verifico filas duplicadas
filas_duplicadas = df_hechos2.duplicated().sum()

if filas_duplicadas > 0:
    print(f"Hay {filas_duplicadas} filas duplicadas.")
else:
    print("No hay filas duplicadas.")



No hay IDs duplicados.
No hay filas duplicadas.


**Victimas**

In [931]:
# Verifico ID duplicados
ids_duplicados_v = df_victimas2['ID_hecho'].duplicated().sum()

if ids_duplicados_v > 0:
    print(f"Hay {ids_duplicados_v} IDs duplicados.")
else:
    print("No hay IDs duplicados.")

# Verifico filas duplicadas
filas_duplicadas_v = df_victimas2.duplicated().sum()

if filas_duplicadas_v > 0:
    print(f"Hay {filas_duplicadas_v} filas duplicadas.")
else:
    print("No hay filas duplicadas.")

Hay 21 IDs duplicados.
No hay filas duplicadas.


De esta información se desprende que los id están duplicados pero no así ninguna fila, probablemente se deba a que son ID de los hechos y estos tienen en ocasiones victimas múltiples. Pero para ello chequeamos primero.

In [932]:
# Agrupar por ID_hecho y contar la cantidad de víctimas
agrupados = df_victimas2.groupby('ID_hecho').size().reset_index(name='N_VICTIMAS')

# Mostrar los accidentes con más de una víctima
accidentes_multiples = agrupados[agrupados['N_VICTIMAS'] > 1]
print(accidentes_multiples)


      ID_hecho  N_VICTIMAS
29   2016-0041           2
97   2016-0126           2
161  2017-0026           2
170  2017-0035           3
171  2017-0036           2
180  2017-0050           2
230  2017-0108           2
234  2017-0112           2
244  2017-0126           2
267  2017-0153           2
289  2018-0015           2
300  2018-0026           2
312  2018-0038           2
359  2018-0085           2
372  2018-0098           2
402  2018-0128           2
427  2019-0010           2
558  2020-0038           2
583  2020-0063           2
596  2020-0077           2


Aquí confirmamos que sí se trata de los accidentes con más de una victima fatal.

### CALIDAD DE DATOS

Empezamos analizando con qué tipo de datos trabajamos para modificar aquellos datos necesarios y a su vez trabajaremos sobre los datos "SD" y otras particularidades que puedan surgir.

**Hechos**

In [933]:
df_hechos2.dtypes

ID                 object
N_VICTIMAS          int64
FECHA              object
AAAA                int64
MM                  int64
DD                  int64
HH                 object
LUGAR_DEL_HECHO    object
TIPO_DE_CALLE      object
Calle              object
Cruce              object
COMUNA              int64
pos x              object
pos y              object
PARTICIPANTES      object
VICTIMA            object
ACUSADO            object
dtype: object

**Fecha**: Se trata de un dato que deberia figurar como fecha y sin embargo lo hace como object. Así que debemos convertirlo.

In [934]:
# Convertir la fecha a formato datetime
#df_hechos2['FECHA'] = pd.to_datetime(df_hechos2['FECHA'])

# Verificar la conversión
#print(df_hechos2['FECHA'].head())
#print(df_hechos2.dtypes)

import warnings
from pandas.errors import SettingWithCopyWarning

# Ignorar advertencias temporalmente
with warnings.catch_warnings():
    warnings.simplefilter("ignore", category=SettingWithCopyWarning)
    
    # Convertir la fecha a formato datetime
    df_hechos2['FECHA'] = pd.to_datetime(df_hechos2['FECHA'], errors='coerce')
        
# Verificar tipos de datos
print(df_hechos2.dtypes)


ID                         object
N_VICTIMAS                  int64
FECHA              datetime64[ns]
AAAA                        int64
MM                          int64
DD                          int64
HH                         object
LUGAR_DEL_HECHO            object
TIPO_DE_CALLE              object
Calle                      object
Cruce                      object
COMUNA                      int64
pos x                      object
pos y                      object
PARTICIPANTES              object
VICTIMA                    object
ACUSADO                    object
dtype: object


Nota: Para evitar la advertencia SettingWithCopyWarning de Pandas, he ajustado el código siguiendo las recomendaciones proporcionadas. Esta advertencia no afecta el análisis actual, pero se maneja para asegurar un análisis limpio y sin interrupciones.

**HH**, al igual que con fecha figura como object y no lo es, así que tambien modficamos el tipo de dato.

In [935]:
# Resumen de valores nulos en cada columna
#nulos2 = df_hechos2.isnull().sum()
#print(nulos2)
sin_datos2 = (df_hechos2 == 'SD').sum()
print(sin_datos2)

ID                  0
N_VICTIMAS          0
FECHA               0
AAAA                0
MM                  0
DD                  0
HH                  1
LUGAR_DEL_HECHO     1
TIPO_DE_CALLE       0
Calle               0
Cruce               0
COMUNA              0
pos x               0
pos y               0
PARTICIPANTES       0
VICTIMA             9
ACUSADO            23
dtype: int64


In [936]:
with warnings.catch_warnings():
    warnings.simplefilter("ignore", category=SettingWithCopyWarning)
    
    # Reemplazar "SD" en HH con 0
    df_hechos2.loc[:, 'HH'] = df_hechos2['HH'].replace('SD', 0).astype(int)
    
    # Convertir HH a int64
    df_hechos2['HH'] = df_hechos2['HH'].astype(int)    

# Verificar tipos de datos y que los nulos se hayan manejado correctamente
print(df_hechos2.dtypes)

ID                         object
N_VICTIMAS                  int64
FECHA              datetime64[ns]
AAAA                        int64
MM                          int64
DD                          int64
HH                          int64
LUGAR_DEL_HECHO            object
TIPO_DE_CALLE              object
Calle                      object
Cruce                      object
COMUNA                      int64
pos x                      object
pos y                      object
PARTICIPANTES              object
VICTIMA                    object
ACUSADO                    object
dtype: object


Nota: Para evitar la advertencia SettingWithCopyWarning de Pandas, he ajustado el código siguiendo las recomendaciones proporcionadas. Esta advertencia no afecta el análisis actual, pero se maneja para asegurar un análisis limpio y sin interrupciones.

**pos x y pos y**

Los pasamos a float

In [937]:
# Ignorar advertencias temporalmente
with warnings.catch_warnings():
    warnings.simplefilter("ignore", category=SettingWithCopyWarning)
    
     # Reemplazar "." con NaN en pos x y pos y usando .loc
    df_hechos2.loc[df_hechos2['pos x'] == '.', 'pos x'] = np.nan
    df_hechos2.loc[df_hechos2['pos y'] == '.', 'pos y'] = np.nan
    
    # Imputar manualmente los valores SD en pos x
    df_hechos2.loc[df_hechos2['ID'] == '2016-0052', 'pos y'] = -34.5482752732176
    df_hechos2.loc[df_hechos2['ID'] == '2016-0136', 'pos y'] = -34.76076916137128
    df_hechos2.loc[df_hechos2['ID'] == '2016-0174', 'pos y'] = -34.620585791298076
    df_hechos2.loc[df_hechos2['ID'] == '2017-0042', 'pos y'] = -34.5482751372226
    df_hechos2.loc[df_hechos2['ID'] == '2017-0050', 'pos y'] = -34.638784958183926
    df_hechos2.loc[df_hechos2['ID'] == '2017-0051', 'pos y'] = -34.66900650255661
    df_hechos2.loc[df_hechos2['ID'] == '2017-0140', 'pos y'] = -34.63587371605594
    df_hechos2.loc[df_hechos2['ID'] == '2018-0039', 'pos y'] = -34.548292810803375
    df_hechos2.loc[df_hechos2['ID'] == '2020-0026', 'pos y'] = -34.54988785980116
    df_hechos2.loc[df_hechos2['ID'] == '2020-0039', 'pos y'] = -34.663132207804004
    df_hechos2.loc[df_hechos2['ID'] == '2021-0023', 'pos y'] = -34.667493
     
    # Imputar nulos en pos y
    df_hechos2.loc[df_hechos2['ID'] == '2016-0052', 'pos x'] = -58.43851883576026
    df_hechos2.loc[df_hechos2['ID'] == '2016-0136', 'pos x'] = -58.180692430054414
    df_hechos2.loc[df_hechos2['ID'] == '2016-0174', 'pos x'] = -58.40123429186223
    df_hechos2.loc[df_hechos2['ID'] == '2017-0042', 'pos x'] = -58.43856363191393
    df_hechos2.loc[df_hechos2['ID'] == '2017-0050', 'pos x'] = -58.49383818773086
    df_hechos2.loc[df_hechos2['ID'] == '2017-0051', 'pos x'] = -58.47555667423749
    df_hechos2.loc[df_hechos2['ID'] == '2017-0140', 'pos x'] = -58.37713882328588
    df_hechos2.loc[df_hechos2['ID'] == '2018-0039', 'pos x'] = -58.43855290307802
    df_hechos2.loc[df_hechos2['ID'] == '2020-0026', 'pos x'] = -58.43541794187507
    df_hechos2.loc[df_hechos2['ID'] == '2020-0039', 'pos x'] = -58.493177387730015
    df_hechos2.loc[df_hechos2['ID'] == '2021-0023', 'pos x'] = -58.419416

    # Convertir a float las dos columnas
    df_hechos2['pos x'] = df_hechos2['pos x'].astype(float)
    df_hechos2['pos y'] = df_hechos2['pos y'].astype(float)
    
    # Verificar la conversión y orden de pos x y pos y
print(df_hechos2[['ID', 'pos x', 'pos y']].head())


          ID      pos x      pos y
0  2016-0001 -58.475340 -34.687570
1  2016-0002 -58.508775 -34.669777
2  2016-0003 -58.390403 -34.631894
3  2016-0004 -58.465039 -34.680930
4  2016-0005 -58.387183 -34.622466


Nota: Para evitar la advertencia SettingWithCopyWarning de Pandas, he ajustado el código siguiendo las recomendaciones proporcionadas. Esta advertencia no afecta el análisis actual, pero se maneja para asegurar un análisis limpio y sin interrupciones.

**Lugar del hecho**

Dado que tiene un solo valor SD y que en la misma fila faltan la mayoría de los datos, podemos eliminarlo para simplificar la tabla y su lectura. 

In [938]:
# Crear una copia del dataframe para evitar el warning
df_hechos2 = df_hechos2.copy()

# Eliminar filas donde LUGAR_DEL_HECHO es 'SD'
df_hechos2.drop(df_hechos2[df_hechos2['LUGAR_DEL_HECHO'] == 'SD'].index, inplace=True)

**Cruces**

Asumimos que los puntos nulos en esta columna se deben a que no todos los accidentes ocurren en cruces con otra calle, y por ende las renombramos como "SIN_CRUCE"

In [939]:
# Reemplazar valores nulos en la columna 'Cruce' con 'SIN_CRUCE'
df_hechos2.loc[:, 'Cruce'] = df_hechos2['Cruce'].fillna('SIN CRUCE')


In [940]:
# Verificar los cambios
print(df_hechos2.isnull().sum())

ID                 0
N_VICTIMAS         0
FECHA              0
AAAA               0
MM                 0
DD                 0
HH                 0
LUGAR_DEL_HECHO    0
TIPO_DE_CALLE      0
Calle              0
Cruce              0
COMUNA             0
pos x              0
pos y              0
PARTICIPANTES      0
VICTIMA            0
ACUSADO            0
dtype: int64


**Victimas**

In [941]:
df_victimas2.dtypes

ID_hecho    object
ROL         object
SEXO        object
EDAD        object
dtype: object

**Edad**

In [942]:
# Reemplaza 'SD' con un valor temporal para poder convertir de object a int
df_victimas2['EDAD'] = df_victimas2['EDAD'].replace('SD', 200)

# Convierte edad a int
df_victimas2['EDAD'] = df_victimas2['EDAD'].astype(int)

# Calculamos la media de las edades, excluyendo las SD (200)
media_edad = round(df_victimas2[df_victimas2['EDAD'] != 200]['EDAD'].mean(),0)
media_edad

# Reemplazamos los datos SD con la media de las edades para no modificar luego las estadísticas de edad
df_victimas2['EDAD'] = df_victimas2['EDAD'].replace(200, media_edad)

#Verificamos resultado
df_victimas2.dtypes

ID_hecho    object
ROL         object
SEXO        object
EDAD         int64
dtype: object

NORMALIZACION DE LOS DATOS

**Hechos**

**Tipo de calle**

In [943]:
tipo_de_calle_cuenta = df_hechos2['TIPO_DE_CALLE'].value_counts()
print(tipo_de_calle_cuenta)

TIPO_DE_CALLE
AVENIDA      429
CALLE        135
AUTOPISTA     66
GRAL PAZ      65
Name: count, dtype: int64


Notamos que que sale como categoría aparte General Paz. Y luego de una breve búsqueda en los datos oficiales del Gobierno de la Ciudad notamos que, si bien se llama Avenida, es considerada una autopista, así que con ese criterio renombramos sus valores a Autopista.
Link: https://buenosaires.gob.ar/movilidad/gerencia-operativa-de-logistica-urbana/normativa-general/avenida-general-paz#:~:text=La%20Avenida%20General%20Paz%20es,sobre%20el%20R%C3%ADo%20Matanza%2DRiachuelo.

In [944]:
# Cambiamos los registros 'GRAL PAZ' por 'AUTOPISTA'
df_hechos2.loc[df_hechos2['TIPO_DE_CALLE'] == 'GRAL PAZ', 'TIPO_DE_CALLE'] = 'AUTOPISTA'
tipo_de_calle_cuenta2 = df_hechos2['TIPO_DE_CALLE'].value_counts()
print(tipo_de_calle_cuenta2)

TIPO_DE_CALLE
AVENIDA      429
CALLE        135
AUTOPISTA    131
Name: count, dtype: int64


**Comuna**

In [945]:
comuna_detalle = df_hechos2['COMUNA'].value_counts()
print(comuna_detalle)

COMUNA
1     90
4     76
9     73
8     65
7     60
3     45
15    44
13    40
12    37
14    35
11    32
10    29
2     25
5     22
6     21
0      1
Name: count, dtype: int64


Notamos que hay un registro de Comuna con valor 0, y luego de investigar, no existe tal comuna, así que miramos en detalle la fila con dicho valor.

In [946]:
fila_cero = df_hechos2[df_hechos2['COMUNA'] == 0]
print(fila_cero)

            ID  N_VICTIMAS      FECHA  AAAA  MM  DD  HH       LUGAR_DEL_HECHO  \
139  2016-0174           1 2016-12-27  2016  12  27   0  AUTOPISTA 25 DE MAYO   

    TIPO_DE_CALLE                 Calle      Cruce  COMUNA      pos x  \
139     AUTOPISTA  AUTOPISTA 25 DE MAYO  SIN CRUCE       0 -58.401234   

         pos y PARTICIPANTES VICTIMA ACUSADO  
139 -34.620586         SD-SD      SD      SD  


Dado que tiene varios datos relevantes ausentes y que se trata de una única fila, tomo la decisión de eliminarla. 

In [947]:
df_hechos2.drop(df_hechos2[df_hechos2['ID'] == '2016-0174'].index, inplace=True)

**Victima**

In [948]:
victimas_detalles = df_hechos2['VICTIMA'].value_counts()
print(victimas_detalles)

VICTIMA
MOTO           295
PEATON         263
AUTO            83
BICICLETA       29
SD               8
CARGAS           7
PASAJEROS        5
MOVIL            2
OBJETO FIJO      1
PEATON_MOTO      1
Name: count, dtype: int64


EL dato peaton_moto pareciera estar errado, así que vemos en detalle dicho valor

In [949]:
fila_peaton_moto = df_hechos2[df_hechos2['VICTIMA'] == 'PEATON_MOTO']
print(fila_peaton_moto)

            ID  N_VICTIMAS      FECHA  AAAA  MM  DD  HH  \
583  2020-0063           2 2020-12-05  2020  12   5   7   

              LUGAR_DEL_HECHO TIPO_DE_CALLE       Calle        Cruce  COMUNA  \
583  NUEVA YORK Y ALTA GRACIA         CALLE  NUEVA YORK  ALTA GRACIA      11   

         pos x     pos y     PARTICIPANTES      VICTIMA ACUSADO  
583 -58.527838 -34.61047  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO  


Dicho y hecho, probablemente confundieron el valor de participantes y se volvió a cargar ese dato en la columna VICTIMA, cuando debieron cargar únicamente PEATON

In [950]:
df_hechos2.loc[df_hechos2['VICTIMA'] == 'PEATON_MOTO', 'VICTIMA'] = 'PEATON'

In [951]:
# Cambiamos nombre de columnas para normalizar datos y unificar con la segunda tabla
df_hechos2.rename(columns={'pos y': 'latitud'}, inplace=True)
df_hechos2.rename(columns={'pos x': 'longitud'}, inplace=True)
df_hechos2.rename(columns={'ID': 'id'}, inplace=True)
df_hechos2.rename(columns={'N_VICTIMAS': 'cantidad_victimas'}, inplace=True)
df_hechos2.rename(columns={'FECHA': 'fecha'}, inplace=True)
df_hechos2.rename(columns={'AAAA': 'anio'}, inplace=True)
df_hechos2.rename(columns={'MM': 'mes'}, inplace=True)
df_hechos2.rename(columns={'DD': 'dia'}, inplace=True)
df_hechos2.rename(columns={'HH': 'franja_hora'}, inplace=True)
df_hechos2.rename(columns={'LUGAR_DEL_HECHO': 'lugar_hecho'}, inplace=True)
df_hechos2.rename(columns={'TIPO_DE_CALLE': 'tipo_calle'}, inplace=True)
df_hechos2.rename(columns={'Calle': 'calle'}, inplace=True)
df_hechos2.rename(columns={'Cruce': 'cruce'}, inplace=True)
df_hechos2.rename(columns={'COMUNA': 'comuna'}, inplace=True)
df_hechos2.rename(columns={'PARTICIPANTES': 'participantes'}, inplace=True)
df_hechos2.rename(columns={'VICTIMA': 'victima'}, inplace=True)
df_hechos2.rename(columns={'ACUSADO': 'acusado'}, inplace=True)
df_hechos2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 694 entries, 0 to 695
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id                 694 non-null    object        
 1   cantidad_victimas  694 non-null    int64         
 2   fecha              694 non-null    datetime64[ns]
 3   anio               694 non-null    int64         
 4   mes                694 non-null    int64         
 5   dia                694 non-null    int64         
 6   franja_hora        694 non-null    int64         
 7   lugar_hecho        694 non-null    object        
 8   tipo_calle         694 non-null    object        
 9   calle              694 non-null    object        
 10  cruce              694 non-null    object        
 11  comuna             694 non-null    int64         
 12  longitud           694 non-null    float64       
 13  latitud            694 non-null    float64       
 14  participantes  

**Victimas**

In [952]:
# Cambiamos nombre de columnas para normalizar datos y unificar con la primera tabla

df_victimas2.rename(columns={'ID_hecho': 'id'}, inplace=True)
df_victimas2.rename(columns={'ROL': 'rol_victima'}, inplace=True)
df_victimas2.rename(columns={'SEXO': 'sexo_victima'}, inplace=True)
df_victimas2.rename(columns={'EDAD': 'edad_victima'}, inplace=True)
df_victimas2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            717 non-null    object
 1   rol_victima   717 non-null    object
 2   sexo_victima  717 non-null    object
 3   edad_victima  717 non-null    int64 
dtypes: int64(1), object(3)
memory usage: 22.5+ KB


## CARGA

In [953]:
df_hechos2.to_csv('hechos_ETL.csv', index=False)
df_victimas2.to_csv('victimas_ETL.csv', index=False)

# COMUNAS

In [954]:
df_comunas = pd.read_csv("comunas.csv", encoding='latin1')
df_comunas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         15 non-null     int64  
 1   objeto     15 non-null     object 
 2   comuna     15 non-null     int64  
 3   barrios    15 non-null     object 
 4   perimetro  15 non-null     float64
 5   area       15 non-null     float64
 6   geometry   15 non-null     object 
dtypes: float64(2), int64(2), object(3)
memory usage: 972.0+ bytes


In [955]:
df_comunas.drop(['id', 'objeto', 'perimetro', 'area', 'geometry'], axis=1, inplace=True)
df_comunas.head()

comuna                                            barrios
0       1  CONSTITUCION, MONSERRAT, RETIRO, PUERTO MADERO...
1       2                                           RECOLETA
2       3                           SAN CRISTOBAL, BALVANERA
3       4    BARRACAS, BOCA, PARQUE PATRICIOS, NUEVA POMPEYA
4       5                                     ALMAGRO, BOEDO

In [956]:
df_comunas.to_csv('comunas_ETL.csv', index=False, encoding = 'latin1')